### ISSUE TITLE
Remote control traffic lights via TraCI

### ISSUE THREAD

**Issue by sotte** [flow-project/flow#529]
**Choose an option**
This is a: 
- [x] Question
- [ ] Feature request
- [ ] Bug report

**Describe the issue**
I want to control the traffic lights with keyboard inputs via TraCI.

Let's assume I run (a slightly changed version of) `examples/sumo/grid.py`. With a hypothetical `remote_control.py` I want to control the traffic lights.
According to the [TraC PythonI](https://sumo.dlr.de/wiki/TraCI/Interfacing_TraCI_from_Python) documentation this should be easy, but flow takes care of starting sumo, and I'm not sure if I can connect to sumo to issue commands.

Am I going into the right direction or is there a better way to control the traffic lights via keyboard input?

**Comment by eugenevinitsky** [comment]
Hi @sotte, just to make sure I'm understanding correctly, you want to make it so you can control a traffic light switch via a keyboard press? If so, this isn't something we've built out but I actually happen to have a branch that shows how to do it; not for traffic lights but for cars. If you look at this branch: https://github.com/flow-project/flow/tree/user_control you should be able to modify the KeyboardController class to change the keys for a traffic light instead of applying acceleration for a car. Hope that helps!

**Comment by eugenevinitsky** [comment]
I should add that this is a very neat thing to have for tutorials, and if you do happen to develop a way to control traffic lights by key press, we would love to merge it in.

**Comment by sotte** [comment]
@eugenevinitsky exactly, I want to control the traffic light via a keyboard press (or a car, or anything really). I have a hacky version of a custom acceleration controller that listens to key presses in a sep. thread and determines the acceleration of a car (directly returns the acceleration without using TraCI). I was about to do the same for the traffic lights, but the structure for traffic lights is different and my implementation felt hacky, so I was looking for a cleaner way via TraCI.

I agree, having this for tutorials or demos is very nice :)

Thanks for pointing out the branch! It looks interesting. Here is the relevant commit:
https://github.com/flow-project/flow/commit/dd9ed6326d16e28d762014afcf03b54e8969e307#diff-358598a49f086ce0f619ece3d53f900c

I'll check out if/how I can adjust that to my needs and get back to you! Thanks!

Just to double check: is it possible to connect via a TraCI script to a running Sumo simulation and change arbitrary things in that simulation? If so, this approach might be a more general solution.

**Comment by eugenevinitsky** [comment]
Hi Stefan, this might suggest that we need a refactor of the traffic lights class to make it simpler to use since it’s easy to do this for cars but hard for traffic lights. Do you have thoughts on what this might look like since you’ve messed around with it a bit?

Also, yes, there is a way to connect TrACI scripts. There is a SumoParam called num_clients that causes SUMO to block until it receives a response from both clients. The second client to connect can override the traci commands of the first. It’s a bit hacky, but might work for now while we figure out a general purpose solution?

**Comment by sotte** [comment]
Hey Eugene,

I haven't figured out a clean way/a way that makes me happy to control the traffic lights, but I'll keep you posted. Thanks for pointing out the TraCI setup!

If you're going to change the interface for traffic lights, I think the interface should be similar to vehicles and their controllers. Maybe something like this:
```python
# This is how it currently looks for cars:
vehicles.add(
    veh_id='rl',
    acceleration_controller=(RLController, {}),
    # ...
)

# This is what I would expect for traffic lights:
tl_logic = TrafficLightParams()
tf_logic.add(
    traffic_light_id1,
    controller=(StaticTrafficLightController, {"phases": phasese}),
)
tf_logic.add(
    traffic_light_id4,
    controller=(BaselineTrafficLightController, {}),
)
tf_logic.add(
    traffic_light_id2,
    controller=(RLTrafficLightController, {}),
)
tf_logic.add(
    traffic_light_id3,
    controller=(KeyboardTrafficLightController, {}),
)

```

Maybe this separates the logic for `Actuated` and `Static` traffic lights a bit.

Keep in mind I'm mostly a user of flow right now, I'm not too familiar with the internals yet! ;)

**Comment by AboudyKreidieh** [comment]
hi @sotte thanks for the suggestion! I definitely like the idea of separating the traffic light controllers the same way we did for the vehicle controllers. @kjang96 think you can start a perhaps detailed issue about how we can go about this? We can then consider assigning it to someone in future releases.

**Comment by eugenevinitsky** [comment]
Hi @sotte, just following up, were you able to get this to work?

**Comment by sotte** [comment]
Hey @eugenevinitsky, due to time constraints we only ended up using the hacky remote control for the cars, not the traffic lights. That being said, if there was a traffic light remote control, we would use it pretty much right away :)

Sadly I also didn't have time to try out the TRACI approach.

I hope that helps!
